src/picongpu/include/particles/ionization/ionization.hpp

```diff
@@ -0,0 +1,386 @@
+        frameIonizer(*ionFrame, linearThreadIdx, cachedB, cachedE, newElectrons);
+        }
+        __syncthreads();
+        /* always true while-loop over all source frames */
+        while (true)
+        {
+            /* < INIT >
+             * - electronId is initialized as -1 (meaning: invalid)
+             * - (local) oldFrameFillLvl set equal to (shared) newFrameFillLvl for each thread 
+             * --> each thread remembers the old "counter"
+             * - then sync */
+            electronId = -1;
+            oldFrameFillLvl = newFrameFillLvl;
+            __syncthreads();
+            /* < ASK >
+             * - if a thread wants to create electrons in each cycle it can do that only once 
+             * and before that it atomically adds to the shared counter and uses the current
+             * value as electronId in the new frame 
+             * - then sync */
+            if (newElectrons > 0) 
+            {
+                electronId = atomicAdd(&newFrameFillLvl, 1);
+            }
+            __syncthreads();
+            /* < EXIT? > 
+             * - if the counter hasn't changed all threads break out of the loop */
+            if (oldFrameFillLvl == newFrameFillLvl)
+            {
+                break;
+            }
+            __syncthreads();
+            /* < FIRST NEW FRAME >
+             * - if there is no frame, yet, the master will create a new target electron frame
+             * and attach it to the back of the frame list 
+             * - sync all threads again for them to know which frame to use */
+            if (linearThreadIdx == 0)
+            {
+                if (electronFrame == NULL)
+                {
+                    electronFrame = &(electronBox.getEmptyFrame());
+                    electronBox.setAsLastFrame(*electronFrame, block);
+                }
+            }
+            __syncthreads();
+            /* < CREATE 1 >
+             * - all electrons fitting into the current frame are created there 
+             * - internal ionization counter is decremented by 1 
+             * - sync */
+            if ((0 <= electronId) && (electronId < maxParticlesInFrame))
+            {
+                /* each thread makes the attributes of its ion accessible */
+                PMACC_AUTO(parentIon,((*ionFrame)[linearThreadIdx]));
+                /* each thread initializes an electron if one should be created */
+                PMACC_AUTO(targetElectronFull,((*electronFrame)[electronId]));
+                
+                /* create an electron in the new electron frame:
+                 * - see particles/ionization/ionizationMethods.hpp */
+                WriteElectronIntoFrame writeElectron;
+                writeElectron(parentIon,targetElectronFull);
+                
+                newElectrons -= 1;
+            }
+            __syncthreads();
+            /* < SECOND NEW FRAME > 
```

Comment:
I somehow feel that you asked that question some weeks ago in a comment on this PR. 
The new electrons ALWAYS fit in two frames because the frame sizes of ions and electrons are the same and each ion-thread is only allowed to create ONE electron at a time. The routine still supports multiple ionization during the same time step. It's just that you get new frames during each run-through of the inner loop. :smile: